app/src/main/java/ru/yandex/architectureproject/presentation/viewmodel/TaskViewModel.kt

```diff
@@ -0,0 +1,68 @@
+package ru.yandex.architectureproject.presentation.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import ru.yandex.architectureproject.domain.AddTaskUseCase
+import ru.yandex.architectureproject.domain.CompleteTaskUseCase
+import ru.yandex.architectureproject.domain.DeleteTaskUseCase
+import ru.yandex.architectureproject.domain.GetAllTasksUseCase
+import ru.yandex.architectureproject.domain.IncompleteTaskUseCase
+import ru.yandex.architectureproject.presentation.state.TaskAction
+import ru.yandex.architectureproject.presentation.state.TaskState
+
+class TaskViewModel(
+    private val addTaskUseCase: AddTaskUseCase,
+    private val deleteTaskUseCase: DeleteTaskUseCase,
+    private val getAllTasksUseCase: GetAllTasksUseCase,
+    private val completeTaskUseCase: CompleteTaskUseCase,
+    private val incompleteTaskUseCase: IncompleteTaskUseCase,
+    private val ioDispatcher: CoroutineDispatcher,
+) : ViewModel() {
+    private val _state = MutableStateFlow<TaskState>(TaskState.Loading)
+    val state: StateFlow<TaskState> = _state.asStateFlow()
+    val taskDeletetingJobsMap = mutableMapOf<Int, Job>()
+
+    init {
+        reduce(TaskAction.LoadingTask)
+    }
+
+    fun reduce(action: TaskAction) {
+        viewModelScope.launch(ioDispatcher) {
+            when (action) {
+                is TaskAction.LoadingTask -> loadTasks()
+                is TaskAction.AddTasks -> addTaskUseCase.invoke(action.task.text)
+                is TaskAction.UpdateTaskStatus -> if (action.isCompleted) {
+                    taskDeletetingJobsMap.put(action.taskID, this.coroutineContext.job)
+                    completeTaskUseCase(action.taskID)
+                } else {
+                    taskDeletetingJobsMap[action.taskID]?.cancel()
```

Comment:
Помимо отмены здесь необходимо еще и удалить джобу из мапы, чтобы она могла быть вычищена из памяти:
```suggestion
                    taskDeletetingJobsMap?.remove(action.taskID)?.cancel()
```